File: tam_agent/analysis/competitive.py
"""Competitive landscape analysis."""

from __future__ import annotations

from tam_agent.llm.client import LLMClient
from tam_agent.llm.prompts import COMPETITIVE_LANDSCAPE, SYSTEM_PROMPT
from tam_agent.models import (
    CompanyInput,
    CompetitiveLandscape,
    IndustryClassification,
    MarketDataPoint,
    SearchResult,
)
from tam_agent.research.data_extraction import extract_market_data, format_search_data_section
from tam_agent.research.web_search import WebSearcher


def analyze_competitive_landscape(
    llm: LLMClient,
    searcher: WebSearcher,
    company: CompanyInput,
    industry: IndustryClassification,
) -> tuple[CompetitiveLandscape, list[SearchResult], list[MarketDataPoint]]:
    """Analyze the competitive landscape for a company's market."""

    context = (
        f"Competitors of {company.name} in {company.industry}, "
        f"NAICS {industry.naics_code}, Geography: {company.geography}"
    )
    queries = llm.generate_search_queries(
        f"Find top competitors, their revenues, and market share data.\n{context}",
        num_queries=4,
    )

    search_results = searcher.multi_search(queries)
    data_points = extract_market_data(llm, search_results, context)
    search_data_section = format_search_data_section(search_results, data_points)

    prompt = COMPETITIVE_LANDSCAPE.format(
        name=company.name,
        industry=company.industry,
        naics_code=industry.naics_code,
        geography=company.geography,
        sub_segments=", ".join(industry.sub_segments),
        search_data_section=search_data_section,
    )

    landscape = llm.structured_query(
        prompt=prompt,
        response_model=CompetitiveLandscape,
        system=SYSTEM_PROMPT,
    )
    return landscape, search_results, data_points

File: src/analysis/competitive.ts
/** Competitive landscape analysis. */

import type { LLMClient } from "../llm/client.js";
import { competitiveLandscapePrompt, SYSTEM_PROMPT } from "../llm/prompts.js";
import {
  CompetitiveLandscapeSchema,
  type CompanyInput,
  type CompetitiveLandscape,
  type IndustryClassification,
  type MarketDataPoint,
  type SearchResult,
} from "../models.js";
import { extractMarketData, formatSearchDataSection } from "../research/data-extraction.js";
import type { WebSearcher } from "../research/web-search.js";

export async function analyzeCompetitiveLandscape(
  llm: LLMClient,
  searcher: WebSearcher,
  company: CompanyInput,
  industry: IndustryClassification,
): Promise<{
  landscape: CompetitiveLandscape;
  searchResults: SearchResult[];
  dataPoints: MarketDataPoint[];
}> {
  const context =
    `Competitors of ${company.name} in ${company.industry}, ` +
    `NAICS ${industry.naics_code}, Geography: ${company.geography}`;

  const queries = await llm.generateSearchQueries(
    `Find top competitors, their revenues, and market share data.\n${context}`,
    4,
  );

  const searchResults = await searcher.multiSearch(queries);
  const dataPoints = await extractMarketData(llm, searchResults, context);
  const searchDataSection = formatSearchDataSection(searchResults, dataPoints);

  const landscape = await llm.structuredQuery(
    competitiveLandscapePrompt({
      name: company.name,
      industry: company.industry,
      naics_code: industry.naics_code,
      geography: company.geography,
      sub_segments: industry.sub_segments.join(", "),
      search_data_section: searchDataSection,
    }),
    CompetitiveLandscapeSchema,
    "CompetitiveLandscape",
    SYSTEM_PROMPT,
  );

  return { landscape, searchResults, dataPoints };
}
